src/mysql.erl

```diff
@@ -712,19 +712,30 @@ encode(Conn, Term) ->
 
 %% @doc Executes the given queries and prepares the given statements after a
 %% connection has been made.
+%%
+%% If any of the queries or prepares fails (ie, returns an `error' tuple),
+%% the connection is closed and an `exit' exception is raised.
 -spec execute_after_connect(connection(), [iodata()], [{atom(), iodata()}])
     -> ok.
 execute_after_connect(Conn, Queries, Prepares) ->
     lists:foreach(fun (Query) ->
                       case query(Conn, Query) of
                           ok -> ok;
                           {ok, _} -> ok;
-                          {ok, _, _} -> ok
+                          {ok, _, _} -> ok;
+                          {error, Reason} ->
+                              stop(Conn),
+                              exit(Reason)
```

Comment:
What's the reasoning for raising `exit` vs `error`?

What if the query raises an exception (e.g. tcp error or badarg such as if Query isn't iodata), is that worth catching?

Reply:
> What's the reasoning for raising `exit` vs `error`?

Hm, I usually do that "by feeling". I have been searching if there was a rule for when to use which, but even a post by Robert Virding didn't shed much light on it, except that `exit` should be used to hint to the calling process that it may want to exit itself. So my thinking was, when a process wants a database connection, it usually _needs_ it, ie can't continue without it.

I can of course change it if you would prefer `error` here.

> What if the query raises an exception (e.g. tcp error or badarg such as if Query isn't iodata), is that worth catching?

Good point, I only thought of a query/prepare being invalid in terms of "invalid for the database", not for `mysql-otp`. I'll add a try around the whole thing then.

Reply:
I think `exit` is suitable for an error you can't recover from by catching it or something like that. If a query crashes an existing connection process, it's an obvious reason for exit, but in start_link, I don't know. Isn't it possible to recover by calling start_link again? From the caller's point of view, it's looks like we just failed to create the connection process.

If we go for exit, it should probably be an exit the try-catch case as well, regardless of what kind of exception was raised because the result is that we stop the connection here too.

Regarding the the Reason, to me it's not obvious that reusing error reasons as exit reasons is the right thing to do. Is it? Another possibility is to raise a general message such as failed to start server and possibly also log the SQL error using logger (or error_logger for older OTPs).

A third possibility is to re-raise the same exception again using the same stacktrace (using erlang:raise/3) after stopping the connection in the catch clause and let errors in query or prepare just raise a case_clause error (i.e. just not adding the `{error, Reason}` case clauses for these).

I'm not sure which one I prefer. Maybe the third one. What do you think?

Reply:
Sounds reasonable to me :)

Reply:
Ok, have a look please :)